File: sales_meet/static/src/js/mdm_dashboard.js
odoo.define('sales_meet.MdmDashboard', function (require) {
"use strict";

var ajax = require('web.ajax');
var ControlPanelMixin = require('web.ControlPanelMixin');
var core = require('web.core');
var Dialog = require('web.Dialog');
var Model = require('web.Model');
var session = require('web.session');
var utils = require('web.utils');
var web_client = require('web.web_client');
var Widget = require('web.Widget');
var _t = core._t;
var QWeb = core.qweb;


var formats = require('web.formats');
var KanbanView = require('web_kanban.KanbanView');
var KanbanRecord = require('web_kanban.Record');
var ActionManager = require('web.ActionManager');

var _lt = core._lt;

var MdmDashboard = Widget.extend(ControlPanelMixin, {
    template: "sales_meet.MdmDashboardMain",
    events: {
        'click .customer_count':'customer_count',
        'click .vendor_count':'vendor_count',
        'click .approved_customer_count':'approved_customer_count',
        'click .approved_vendor_count':'approved_vendor_count',
        'click .created_customer_count':'created_customer_count',
        'click .created_vendor_count':'created_vendor_count',
        'click .product_count':'product_count',
    },

    init: function(parent, context) {
        this._super(parent, context);
        var self = this;
        login_user = true;
        this._super(parent,context);

    },

    start: function() {
        var self = this;

        var model  = new Model('res.partner').call('get_mdm_details').then(function(result){
            this.login_user =  result[0];
            // self.login_user =  result[0]

            $('.o_mdm_dashboard').html(QWeb.render('MdmManagerDashboard', {widget: this}));

        });

    },


    customer_count: function(e) {
        var self = this;
        e.stopPropagation();
        e.preventDefault();
        this.do_action({
            name: _t("Customer"),
            type: 'ir.actions.act_window',
            res_model: 'res.partner',
            view_mode: 'tree,form',
            view_type: 'form',
            views: [[false, 'list'],[false, 'form']],
            domain: [['customer','=', true],['active','=', false],['state', 'in', ['draft','Submitted']]],

            target: 'current'
        })
    },

    approved_customer_count: function(e) {
        var self = this;
        e.stopPropagation();
        e.preventDefault();
        this.do_action({
            name: _t("Approved Customer"),
            type: 'ir.actions.act_window',
            res_model: 'res.partner',
            view_mode: 'tree,form',
            view_type: 'form',
            views: [[false, 'list'],[false, 'form']],
            domain: [['customer','=', true],['active','=', false],['state', '=', 'Approved']],

            target: 'current'
        })
    },

    created_customer_count: function(e) {
        var self = this;
        e.stopPropagation();
        e.preventDefault();
        this.do_action({
            name: _t("Customer"),
            type: 'ir.actions.act_window',
            res_model: 'res.partner',
            view_mode: 'tree,form',
            view_type: 'form',
            views: [[false, 'list'],[false, 'form']],
            domain: [['customer','=', true],['active','=', true],['state', 'in', ['created','updated']]],

            target: 'current'
        })
    },

    vendor_count: function(e) {
        var self = this;
        e.stopPropagation();
        e.preventDefault();
        this.do_action({
            name: _t("Vendor"),
            type: 'ir.actions.act_window',
            res_model: 'res.partner',
            view_mode: 'tree,form',
            view_type: 'form',
            views: [[false, 'list'],[false, 'form']],
            domain: [['supplier','=', true],['active','=', false],['state', 'in', ['draft','Submitted']]],
            target: 'current'
        })
    },

    approved_vendor_count: function(e) {
        var self = this;
        e.stopPropagation();
        e.preventDefault();
        this.do_action({
            name: _t("Vendor"),
            type: 'ir.actions.act_window',
            res_model: 'res.partner',
            view_mode: 'tree,form',
            view_type: 'form',
            views: [[false, 'list'],[false, 'form']],
            domain: [['supplier','=', true],['active','=', false],['state', '=', 'Approved']],
            target: 'current'
        })
    },

    created_vendor_count: function(e) {
        var self = this;
        e.stopPropagation();
        e.preventDefault();
        this.do_action({
            name: _t("Vendor"),
            type: 'ir.actions.act_window',
            res_model: 'res.partner',
            view_mode: 'tree,form',
            view_type: 'form',
            views: [[false, 'list'],[false, 'form']],
            domain: [['supplier','=', true],['active','=', true],['state', 'in', ['created','updated']]],
            target: 'current'
        })
    },

    product_count: function(e) {
        var self = this;
        e.stopPropagation();
        e.preventDefault();
        var date = new Date();
        this.do_action({
            name: _t("Product"),
            type: 'ir.actions.act_window',
            res_model: 'product.product',
            view_mode: 'tree,form',
            view_type: 'form',
            views: [[false, 'list'],[false, 'form']],
            domain: [['active','=', false]],
            target: 'current'
        })
    },
   

});

core.action_registry.add('mdm_dashboard', MdmDashboard);

return MdmDashboard;

});

File: sales_meet/models/mdm_dashboard.py

 
from odoo import models, api, _
from odoo.http import request
import calendar
from odoo.tools import DEFAULT_SERVER_DATE_FORMAT
from datetime import datetime, timedelta , date
import time

class ResPartnerExtension(models.Model):
    _inherit = 'res.partner'


    def get_attached_docs(self):
        meeting_ids = self.env['calendar.event'].search(['|',("lead_id","=",self.id),("opportunity_id","=",self.id)])
        imd = self.env['ir.model.data']
        action = imd.xmlid_to_object('sales_meet.action_calendar_event_crm')
        list_view_id = imd.xmlid_to_res_id('sales_meet.view_calendar_event_tree_extension')
        form_view_id = imd.xmlid_to_res_id('sales_meet.view_calendar_event_form_extension')
        result = {
            'name': action.name,
            'help': action.help,
            'type': action.type,
            'views': [[list_view_id, 'tree'], [form_view_id, 'form'],
                [False, 'graph'], [False, 'kanban'], [False, 'pivot']],
            'target': action.target,
            'context': action.context,
            'res_model': action.res_model,
        }
        if len(meeting_ids) > 1:
            result['domain'] = "[('id','in',%s)]" % meeting_ids.ids
        elif len(meeting_ids) == 1:
            result['views'] = [(form_view_id, 'form')]
            result['res_id'] = meeting_ids.ids[0]
        else:
            result = {'type': 'ir.actions.act_window_close'}
        return result


    def get_mdm_details(self):
        uid = request.session.uid
        cr = self.env.cr

        user_id = self.env['res.users'].sudo().search_read([('id', '=', uid)], limit=1)
              
        date_today = datetime.today()
        date_from = datetime.today().replace(day=1)
        date_to = datetime.now().replace(day = calendar.monthrange(datetime.now().year, datetime.now().month)[1])

        customer_count = self.env['res.partner'].sudo().search_count([('active', '=', False),
                                                        ('state', 'in', ('draft','Submitted')),
                                                        ('customer','=',True)])

        approved_customer_count = self.env['res.partner'].sudo().search_count([('active', '=', False),
                                                        ('state', '=', 'Approved'),
                                                        ('customer','=',True)])

        created_customer_count = self.env['res.partner'].sudo().search_count([('active', '=', True),
                                                        ('state', 'in', ('created','updated')),
                                                        ('customer','=',True)])

        vendor_count = self.env['res.partner'].sudo().search_count([('active', '=', False),
                                                        ('state', 'in', ('draft','Submitted')),
                                                        ('supplier','=',True)])

        approved_vendor_count = self.env['res.partner'].sudo().search_count([('active', '=', False),
                                                        ('state', '=', 'Approved'),
                                                        ('supplier','=',True)])
        created_vendor_count = self.env['res.partner'].sudo().search_count([('active', '=', True),
                                                        ('state', 'in', ('created','updated')),
                                                        ('supplier','=',True)])


        product_count = self.env['product.product'].sudo().search_count([('active', '=', False)])

        if user_id:
            data = {
                'customer_count': customer_count,
                'vendor_count': vendor_count,
                'approved_customer_count': approved_customer_count,
                'approved_vendor_count': approved_vendor_count,
                'created_customer_count': created_customer_count,
                'created_vendor_count': created_vendor_count,
                'product_count': product_count,                
            }
            user_id[0].update(data)

        return user_id